libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt

```diff
@@ -49,8 +50,13 @@ class ReplayProgressObserver(
         if (routeProgressRouteLeg != null) {
             val replayEvents = replayRouteMapper.mapRouteLegGeometry(routeProgressRouteLeg)
             if (replayEvents.isNotEmpty()) {
+                val wasPlaying = mapboxReplayer.isPlaying()
+                mapboxReplayer.clearEvents()
```

Comment:
this will cause a lot of side effects. it also is not a solution for really long routeLegs

another way to reduce the memory footprint of the replayer would be to put the `ReplayEvents(mutableListOf())` list into a file

Reply:
The ReplayProgressObserver already produces a lot of side effects and controls which samples are being played. If left to its own devices, it would accumulate the events indefinitely, even though most of them are detached from the current context (for example when the route changes to a new one).

My understanding is that if someone uses the ReplayProgressObserver, they are accepting the fact that this class will control the events that the replayer works with.

The problem I see is that even if ReplayProgressObserver does not clear the events automatically, we'd have to communicate to developers when they should clear the events manually in this case to avoid the memory growth in some of the scenarios, but there is no good hook to do that without breaking what ReplayProgressObserver was doing.

> another way to reduce the memory footprint of the replayer would be to put the ReplayEvents(mutableListOf()) list into a file

Could you elaborate? I definitely don't think the changes in the PR are a silver bullet and I'd love to find a better way to keep the number of events generate by the ReplayProgressObserver in check.

Maybe the ReplayProgressObserver should only clean up the events that it added instead of clearing all of them?

Reply:
> Could you elaborate? I definitely don't think the changes in the PR are a silver bullet and I'd love to find a better way to keep the number of events generate by the ReplayProgressObserver in check.

Yeah instead of holding the event in memory they would be queried from a file with a cursor. Interface is all the same, but instead of bloating memory you're using a file. Obviously more complicated, and requires a new dependency to get access to the file system.
```
- val replayEvents = ReplayEvents(mutableListOf())
+ val replayEvents = ReplayEvents(filePath)
```

So I think it would is better to clear the events when the DirectionsRoute changes. This one also "resets" replay to the device location. We're using this implementation in a few places so it's worth adding IMO.
``` kotlin
class ReplayRoutesObserver(
    val mapboxReplayer: MapboxReplayer,
    val context: Context
) : RoutesObserver {
    override fun onRoutesChanged(result: RoutesUpdatedResult) {
        if (result.routes.isEmpty()) {
            mapboxReplayer.clearEvents()
            MapboxNavigationProvider.retrieve().resetTripSession()
            mapboxReplayer.pushRealLocation(context, 0.0)
            mapboxReplayer.play()
        }
    }
}
```

The next issue is that we have to subscribe 2 tools to get basic usage. [There is progress for a component that handles all these things](https://github.com/mapbox/mapbox-navigation-android/blob/05748f2e4a01db1c0fb41a8999a5ee7cd8e69eac/qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/lifecycle/DropInReplayViewModel.kt#L40-L41) in the qa-test-app

Reply:
> The ReplayProgressObserver already produces a lot of side effects and controls which samples are being played. If left to its own devices, it would accumulate the events indefinitely, even though most of them are detached from the current context (for example when the route changes to a new one).

Also agree here. It's mostly a testing tool and your proposal to clear each `RouteLeg` is good. When we're dealing with seekTo issues, then we could solve those issues with another tool.

Sorry that I'm changing my mind, but yeah.. Clearing for each `RouteLeg` has my 💚